File: facility/ik.py
"""
facility/ik.py

Two-link planar IK utilities with degree-based input/output.

Contains:
- two_link_ik_degrees(...)     -- a clearer, robust standard formulation (law of cosines).

The function returns (hip_deg, knee_deg, reachable).
"""
from __future__ import annotations
import math
from typing import Tuple, Optional


def _clamp(x: float, a: float, b: float) -> float:
    return max(a, min(b, x))


def _normalize_angle_rad(x: float) -> float:
    """Normalize radian angle to (-pi, pi]."""
    return (x + math.pi) % (2.0 * math.pi) - math.pi


def two_link_ik_degrees(angle_deg: float,
                        r: float,
                        l1: float,
                        l2: float,
                        elbow_up: Optional[bool] = True,
                        allow_unreachable: bool = True,
                        prev_angles: Optional[Tuple[float, float]] = None
                       ) -> Tuple[float, float, bool]:
    """
    Robust two-link planar IK with degree inputs/outputs using the law of cosines.

    Args:
      angle_deg: target direction in degrees (from +X, CCW positive).
      r: target distance (>= 0).
      l1, l2: link lengths (>0).
      elbow_up:
        - True: choose elbow-up (+acos) solution
        - False: choose elbow-down (-acos)
        - None: if prev_angles provided, pick solution closest to prev_angles; otherwise defaults to True
      allow_unreachable: if True clamp r into reachable range and return reachable=False; if False return reachable=False and no angles.
      prev_angles: optional (hip_deg, knee_deg) to choose continuous solution when elbow_up is None.

    Returns:
      (hip_deg, knee_deg, reachable)
        hip_deg: hip/base angle in degrees
        knee_deg: signed knee angle in degrees (positive = elbow_up direction)
        reachable: True if original r was reachable
    """
    if l1 <= 0.0 or l2 <= 0.0:
        raise ValueError("link lengths must be positive")

    angle = math.radians(angle_deg)

    # handle zero-distance target
    if r <= 0.0:
        # Hip points at desired direction; knee straight
        return angle_deg, 0.0, True

    max_r = l1 + l2
    min_r = abs(l1 - l2)
    reachable = True
    if r > max_r:
        reachable = False
        if allow_unreachable:
            r = max_r
        else:
            return 0.0, 0.0, False
    elif r < min_r:
        reachable = False
        if allow_unreachable:
            r = min_r
        else:
            return 0.0, 0.0, False

    # law of cosines for knee angle
    cos_t2 = (r * r - l1 * l1 - l2 * l2) / (2.0 * l1 * l2)
    cos_t2 = _clamp(cos_t2, -1.0, 1.0)
    t2_abs = math.acos(cos_t2)

    # two candidate knee angles (signed)
    t2_candidates = [t2_abs, -t2_abs]

    # compute corresponding hip angles for each candidate
    solutions = []
    for t2 in t2_candidates:
        s2 = math.sin(t2)
        c2 = math.cos(t2)
        k1 = l1 + l2 * c2
        k2 = l2 * s2
        t1 = angle - math.atan2(k2, k1)
        solutions.append((t1, t2))

    # choose solution based on elbow_up or prev_angles
    chosen = None
    if elbow_up is True:
        chosen = solutions[0]  # +t2_abs as 'elbow up' convention
    elif elbow_up is False:
        chosen = solutions[1]
    else:
        # elbow_up is None => choose solution closest to prev_angles if provided
        if prev_angles is not None:
            prev_hip_rad = math.radians(prev_angles[0])
            prev_knee_rad = math.radians(prev_angles[1])
            best = None
            best_dist = None
            for (t1_cand, t2_cand) in solutions:
                dh = _normalize_angle_rad(t1_cand - prev_hip_rad)
                dk = _normalize_angle_rad(t2_cand - prev_knee_rad)
                dist = dh * dh + dk * dk
                if best is None or dist < best_dist:
                    best = (t1_cand, t2_cand)
                    best_dist = dist
            chosen = best
        else:
            # default to elbow_up True
            chosen = solutions[0]

    t1, t2 = chosen
    hip_deg = math.degrees(t1)
    knee_deg = math.degrees(t2)
    return hip_deg, knee_deg, reachable

File: src/facility/ik.cpp
#include "facility/ik.hpp"

namespace facility {

static double clampd(double x, double a, double b) {
    if (x < a) return a;
    if (x > b) return b;
    return x;
}

static double normalize_angle_rad(double x) {
    const double two_pi = 2.0 * M_PI;
    x = std::fmod(x + M_PI, two_pi);
    if (x < 0) x += two_pi;
    return x - M_PI;
}

IKResult two_link_ik_degrees(double angle_deg,
                             double r,
                             double l1,
                             double l2,
                             int elbow_choice,
                             bool allow_unreachable,
                             const std::optional<std::pair<double,double>>& prev_angles) {
    if (l1 <= 0.0 || l2 <= 0.0) throw std::invalid_argument("link lengths must be positive");

    const double deg2rad = M_PI / 180.0;
    const double rad2deg = 180.0 / M_PI;

    double angle = angle_deg * deg2rad;

    if (r <= 0.0) {
        return IKResult{ angle_deg, 0.0, true };
    }

    double max_r = l1 + l2;
    double min_r = std::fabs(l1 - l2);
    bool reachable = true;
    if (r > max_r) {
        reachable = false;
        if (allow_unreachable) r = max_r;
        else return IKResult{0.0, 0.0, false};
    } else if (r < min_r) {
        reachable = false;
        if (allow_unreachable) r = min_r;
        else return IKResult{0.0, 0.0, false};
    }

    double cos_t2 = (r*r - l1*l1 - l2*l2) / (2.0 * l1 * l2);
    cos_t2 = clampd(cos_t2, -1.0, 1.0);
    double t2_abs = std::acos(cos_t2);

    double t2_cands[2] = { t2_abs, -t2_abs };
    double sols_t1[2];
    double sols_t2[2];

    for (int i = 0; i < 2; ++i) {
        double t2 = t2_cands[i];
        double s2 = std::sin(t2);
        double c2 = std::cos(t2);
        double k1 = l1 + l2 * c2;
        double k2 = l2 * s2;
        double t1 = angle - std::atan2(k2, k1);
        sols_t1[i] = t1;
        sols_t2[i] = t2;
    }

    int chosen_idx = 0;
    if (elbow_choice == 1) chosen_idx = 0;
    else if (elbow_choice == -1) chosen_idx = 1;
    else {
        // choose based on prev_angles if provided, otherwise default to elbow-up (index 0)
        if (prev_angles.has_value()) {
            double ph = prev_angles->first * deg2rad;
            double pk = prev_angles->second * deg2rad;
            double bestd = std::numeric_limits<double>::infinity();
            for (int i = 0; i < 2; ++i) {
                double dh = normalize_angle_rad(sols_t1[i] - ph);
                double dk = normalize_angle_rad(sols_t2[i] - pk);
                double d = dh*dh + dk*dk;
                if (d < bestd) { bestd = d; chosen_idx = i; }
            }
        } else {
            chosen_idx = 0;
        }
    }

    double hip_deg = sols_t1[chosen_idx] * rad2deg;
    double knee_deg = sols_t2[chosen_idx] * rad2deg;
    return IKResult{ hip_deg, knee_deg, reachable };
}

} // namespace facility
